pkg/controllers/machine_controller.go

```diff
@@ -285,6 +298,109 @@ func (r *MachineReconciler) startCheckAndEnqueueMachineState(ctx context.Context
 	}, r.resyncIntervalMachineState)
 }
 
+func (r *MachineReconciler) startGarbageCollector(ctx context.Context, log logr.Logger) {
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
+		machines, err := r.machines.List(ctx)
+		if err != nil {
+			log.Error(err, "failed to list machines")
+			return
+		}
+
+		for _, machine := range machines {
+			if !slices.Contains(machine.Finalizers, MachineFinalizer) || machine.DeletedAt == nil {
+				continue
+			}
+
+			logger := log.WithValues("machineID", machine.ID)
+			if err := r.processMachineDeletion(ctx, logger, machine); err != nil {
+				logger.Error(err, "failed to garbage collect machine")
+			}
+		}
+
+	}, r.resyncIntervalGarbageCollector)
+}
+
+func (r *MachineReconciler) processMachineDeletion(ctx context.Context, log logr.Logger, machine *api.Machine) error {
+	isDeleting, err := r.deleteMachine(ctx, log, machine)
+	switch {
+	case isDeleting:
+		return nil
+	case err != nil:
+		return fmt.Errorf("failed to delete machine: %w", err)
+	}
+	log.V(1).Info("Deleted machine")
+
+	if err := r.deleteVolumes(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine disks: %w", err)
+	}
+	log.V(1).Info("Removed machine disks")
+
+	if err := r.deleteNetworkInterfaces(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine network interfaces: %w", err)
+	}
+	log.V(1).Info("Removed network interfaces")
+
+	if err := os.RemoveAll(r.host.MachineDir(machine.ID)); err != nil {
+		return fmt.Errorf("failed to remove machine directory: %w", err)
+	}
+	log.V(1).Info("Removed machine directory")
+
+	machine.Finalizers = utils.DeleteSliceElement(machine.Finalizers, MachineFinalizer)
+	if _, err := r.machines.Update(ctx, machine); store.IgnoreErrNotFound(err) != nil {
+		return fmt.Errorf("failed to update machine metadata: %w", err)
+	}
+	log.V(1).Info("Removed Finalizer. Deletion completed")
+
+	return nil
+}
+
+func (r *MachineReconciler) deleteMachine(ctx context.Context, log logr.Logger, machine *api.Machine) (bool, error) {
+	domain := libvirt.Domain{
+		UUID: libvirtutils.UUIDStringToBytes(machine.ID),
+	}
+
+	if machine.Spec.ShutdownAt.IsZero() {
+		machine.Spec.ShutdownAt = time.Now()
+		if _, err := r.machines.Update(ctx, machine); err != nil {
+			return false, fmt.Errorf("failed to update ShutdownAt: %w", err)
+		}
+		log.V(1).Info("Updated machine ShutdownAt", "ShutdownAt", machine.Spec.ShutdownAt)
+	}
+
+	if time.Now().Before(machine.Spec.ShutdownAt.Add(r.gcVMGracefulShutdownTimeout)) {
```

Comment:
Won't we run the shutdown multiple times here until the shutdown time is reached? Is that the expected behavior?

Reply:
yes it is expected behavior because sometimes signal can be ignore and shutdown call is async call.

Reply:
How can a signal be "ignored"? Async should be not a problem here. I was just wondering why we should pounding a shutdown signal to the domain every X seconds.

Reply:
pls ask @gehoern, he recommended call it more times (min 2times).

Reply:
Async isn't problem and it help us here but it could be problem, if it is blocking call as Destroy.

Reply:
The `VM` can miss the `AcpiPowerBtn` signal in case of heavy load. Therefore the signal is sent multiple times. Maybe put a comment in the code to document it.

Reply:
Comment added.

Reply:
Fair enought. Then I would suggest to put some comments in the code explaining this circumstances.

Reply:
I see this has been addressed already.